src/dataframe/dataframe.jl

```diff
@@ -966,23 +966,46 @@ function Base.delete!(df::DataFrame, inds)
     if !isempty(inds) && size(df, 2) == 0
         throw(BoundsError(df, (inds, :)))
     end
+
     # we require ind to be stored and unique like in Base
     # otherwise an error will be thrown and the data frame will get corrupted
-    foreach(col -> deleteat!(col, inds), _columns(df))
-    return df
+    return _delete!_helper(df, inds)
 end
 
 function Base.delete!(df::DataFrame, inds::AbstractVector{Bool})
     if length(inds) != size(df, 1)
         throw(BoundsError(df, (inds, :)))
     end
     drop = findall(inds)
-    foreach(col -> deleteat!(col, drop), _columns(df))
-    return df
+    return _delete!_helper(df, drop)
 end
 
 Base.delete!(df::DataFrame, inds::Not) = delete!(df, axes(df, 1)[inds])
 
+function _delete!_helper(df::DataFrame, drop)
+    cols = _columns(df)
+    isempty(cols) && return df
+
+    n = nrow(df)
+    col1 = cols[1]
+    deleteat!(col1, drop)
+    newn = length(col1)
+
+    for i in 2:length(cols)
+        col = cols[i]
+        if length(col) == n
+            deleteat!(col, drop)
+        end
+    end
+
+    for i in 1:length(cols)
+        # this should never happen, but we add it for safety
+        @assert length(cols[i]) == newn corrupt_msg(df, i)
```

Comment:
can you think of a scenario when this could be triggered (I cannot unless there is some bug in `deleteat!` for some special vector type, but who knows)?

Reply:
That would be a serious bug in a custom array implementation. Probably not worth checking here as it's not particular to data frames.

Reply:
But it is very cheap in comparison to all other things we do. We have found many serious bugs even in Julia Base when developing DataFrames.jl. Do you think it is a problem to have this check? (as you know I prefer to have an overly safe code)

Reply:
If you're willing to write tests that cover these lines, no problem. :-p

Reply:
This is easy:
```
    df = DataFrame(a=1, b=2)
    push!(df.b, 3)
    @test_throws AssertionError delete!(df, 1)
```
I have added this test :smile:.